### ISSUE TITLE
TPROXY iptables-rules for cilium-dns-egress redirect to the wrong port

### ISSUE THREAD

**Issue by miono** [cilium/cilium#36063]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.16.3 and lower than v1.17.0

### What happened?

We are running kubernetes clusters on GKE. Sometimes the cluster-autoscaler backs out of a node scale down if a pod doesn't exit cleanly within the timeout. When this happens the node scaledown fails and the node is put back in to the pool.

Pods getting scheduled on these nodes can't do any DNS-lookups if there is a CNP with an L7 DNS rule for example:
```
spec:
  egress:
  - toEndpoints:
    - matchLabels:
        k8s:io.kubernetes.pod.namespace: kube-system
        k8s:k8s-app: kube-dns
    toPorts:
    - ports:
      - port: "53"
        protocol: ANY
      rules:
        dns:
        - matchPattern: '*'
  endpointSelector:
    matchLabels: {}

```

I've tried tracking the issue down to the best of my abilities and here's what I've come up with:
The DNS-proxy is listening on a dynamically allocated port, in this case 35009:
```
root@gke-us1-us1-default-2022-08-223cced9-74qn:/home/cilium# ss -nltup | grep cilium-agent
udp   UNCONN 0      0           127.0.0.1:35009      0.0.0.0:*    users:(("cilium-agent",pid=1,fd=45))
tcp   LISTEN 0      1024        127.0.0.1:9890       0.0.0.0:*    users:(("cilium-agent",pid=1,fd=7))
tcp   LISTEN 0      1024        127.0.0.1:9878       0.0.0.0:*    users:(("cilium-agent",pid=1,fd=51))
tcp   LISTEN 0      1024        10.1.0.87:4240       0.0.0.0:*    users:(("cilium-agent",pid=1,fd=49))
tcp   LISTEN 0      1024        127.0.0.1:35009      0.0.0.0:*    users:(("cilium-agent",pid=1,fd=39))
tcp   LISTEN 0      1024                *:9091             *:*    users:(("cilium-agent",pid=1,fd=66))
tcp   LISTEN 0      1024                *:9990             *:*    users:(("cilium-agent",pid=1,fd=8))
tcp   LISTEN 0      1024                *:4244             *:*    users:(("cilium-agent",pid=1,fd=77))

```

However the iptables TPROXY-rule is pointing to another port:
```
root@gke-us1-us1-default-2022-08-223cced9-74qn:/home/cilium# iptables -t mangle -nvL | grep cilium-dns-egress
   64  3840 TPROXY     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0xa9330200 /* cilium: TPROXY to host cilium-dns-egress proxy */ TPROXY redirect 127.0.0.1:13225 mark 0x200/0xffffffff
90913 9181K TPROXY     udp  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0xa9330200 /* cilium: TPROXY to host cilium-dns-egress proxy */ TPROXY redirect 127.0.0.1:13225 mark 0x200/0xffffffff

```

In the proxy_ports_state.json file we have the "iptables-port" listed:
```
root@gke-us1-us1-default-2022-08-223cced9-74qn:/home/cilium# cat /var/run/cilium/state/proxy_ports_state.json
{"cilium-dns-egress":{"type":"dns","ingress":false,"port":13225}}
```

There is some weird things going on in the logs that I can't quite understand:
`time="2024-11-13T06:10:58Z" level=info msg="Reusing previous DNS proxy port: 35009" subsys=daemon` This is presumably the port the dnsproxy ran on before the cilium pod got restarted.
`time="2024-11-13T06:11:00Z" level=info msg="Adding new proxy port rules for cilium-dns-egress:35009" id=cilium-dns-egress subsys=proxy` This feels normal to me.
However, a few seconds later we get this in the logs:
`time="2024-11-13T06:11:03Z" level=info msg="Adding new proxy port rules for cilium-dns-egress:0" id=cilium-dns-egress subsys=proxy`
And then:
`time="2024-11-13T06:11:03Z" level=info msg="Adding new proxy port rules for cilium-dns-egress:13225" id=cilium-dns-egress subsys=proxy`

I have done some digging into the code. (caveat, I'm not the worlds best programmer and I'm not entirely familiar with the cilium codebase, but here's what I gathered)
1. The log message "Adding new proxy port rules for..." Comes from `pkg/proxy/proxy.go` on line 247 in the function `*Proxy.ackProxyPort`. This function is only called from the exported function `*Proxy.AckProxyPort`
2. `*Proxy.AckProxyPort` is called from `daemon/cmd/fqdn.go` on line 179 in the function `*Daemon.updateDNSDatapathRules` (It's also called from the `ciliumenvoyconfig` package, but I think that is irrelevant in this case.
3. `*Daemon.updateDNSDatapathRules` is called from `daemon/cmd/daemon.go` on line 871 in the function `newDaemon`.
4. `newDaemon`  is called from `daemon/cmd/daemon_main.go` on line 1740 in the `newDaemonPromise` function
5. This is called(?) through the `cell.Provide` var in `daemon/cmd/daemon_main.go` on line 1633. But this is where it gets too hairy for me to follow due to my limited knowledge.

### How can we reproduce the issue?

We haven't been able to reproduce this ourselves, the only time we see it happening is when a node scale down fails and the node gets reinstated.

### Cilium Version

 miono@laima:~$ cilium -n flux version
cilium-cli: v0.16.20 compiled with go1.23.2 on linux/amd64
cilium image (default): v1.16.3
cilium image (stable): v1.16.3
cilium image (running): 1.16.3

### Kernel Version

Linux gke-us1-us1-default-2022-08-223cced9-74qn 6.1.100+ #1 SMP PREEMPT_DYNAMIC Sat Aug 24 16:19:44 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Client Version: v1.30.6-dispatcher
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.5-gke.1014003

### Regression

We saw these issues appearing after upgrading kubernetes to 1.30 and cilium to 1.16. We had failed node scale downs previously too, without these problems occuring.

### Sysdump

This would leak quite a lot of internal information, please advise if there are some specific part(s) of the sysdump you would need and I can try to scrub the internal data from the files and provide them.

### Relevant log output

```shell
time="2024-11-13T06:10:58Z" level=info msg="Reusing previous DNS proxy port: 35009" subsys=daemon
time="2024-11-13T06:11:00Z" level=info msg="Adding new proxy port rules for cilium-dns-egress:35009" id=cilium-dns-egress subsys=proxy
time="2024-11-13T06:11:03Z" level=info msg="Adding new proxy port rules for cilium-dns-egress:0" id=cilium-dns-egress subsys=proxy
time="2024-11-13T06:11:03Z" level=info msg="Adding new proxy port rules for cilium-dns-egress:13225" id=cilium-dns-egress subsys=proxy
```

### Cilium Users Document

- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by jrajahalme** [comment]
@miono Thanks for the detailed report. Cilium 1.16.4 (released earlier today) adds a new Cilium agent option `--restored-proxy-ports-age-limit` that defaults to 15 minutes. Do you know how long affected nodes are offline before being reinstated?

Note that Cilium agent rewrites `proxy_ports_state.json` file any time a new proxy port is acknowledged, so the file may have been overwritten if you observed it while looking into the problem.

It would be good to know if this issue reproduces with Cilium 1.16.4.

**Comment by miono** [comment]
@jrajahalme The logs from the cluster autoscaler would suggest that it takes about 10-11 minutes before they get reinstated:
`11/21/2024 6:50:57.980 AM +0100 Scale-down: removing node gke-us1-us1-default-2022-08-223cced9-bq97, utilization: {0.23801967629324025 0.13656415510445066 0 cpu 0.23801967629324025}, pods to reschedule: client-777644b885-27tzt,shipments-6579969685-rr2m2`
`11/21/2024 7:01:31.173 AM +0100 failed to drain and delete node: Failed to drain node /gke-us1-us1-default-2022-08-223cced9-bq97: pods remaining after timeout`

I will see if I can upgrade to cilium 1.16.4 in one of our clusters.

Currently we're trying to work around the issue by using the `--tofqdns-proxy-port='5353'"` flag (To force it to always take the same port), does that have any known side effects that you know about?

**Comment by miono** [comment]
@jrajahalme I have an indication from the logs on a node that this has happened also on 1.16.4. Unfortunately the node got successfully scaled down before I got the chance to investigate further, but the logs suggest the same type of issue as in 1.16.3.
```
time="2024-11-21T16:11:16Z" level=info msg="Reusing previous DNS proxy port: 5353" subsys=daemon
time="2024-11-21T16:11:18Z" level=info msg="Adding new proxy port rules for cilium-dns-egress:5353" id=cilium-dns-egress subsys=proxy
time="2024-11-21T16:11:21Z" level=info msg="Adding new proxy port rules for cilium-dns-egress:0" id=cilium-dns-egress subsys=proxy
time="2024-11-21T16:11:21Z" level=info msg="Adding new proxy port rules for cilium-dns-egress:19357" id=cilium-dns-egress subsys=proxy
```

**Comment by jrajahalme** [comment]
Turns out it was possible for the DNS proxy port to get released when an endpoint with a DNS policy failed to regenerate successfully. This should be fixed by https://github.com/cilium/cilium/pull/36142.

**Comment by miono** [comment]
@jrajahalme Thanks! A couple of questions:

1. Do you think it's probably that this will be included in 1.16.5, and when do you think 1.16.5 can be released?
2. We thought we could work around this by setting a static dnsproxy-port, but it seems like we can end up in the same situation also with dnsproxy.port set (in our case to `5353`):
```
root@gke-us1-us1-default-2022-08-223cced9-cc8k:/home/cilium# cat /tmp/cilium/config-map/tofqdns-proxy-port ; echo
5353
root@gke-us1-us1-default-2022-08-223cced9-cc8k:/home/cilium# ss -nltup | grep cilium
udp   UNCONN 0      0            127.0.0.1:5353       0.0.0.0:*    users:(("cilium-agent",pid=1,fd=46))
tcp   LISTEN 0      1024         127.0.0.1:9878       0.0.0.0:*    users:(("cilium-agent",pid=1,fd=56))
tcp   LISTEN 0      1024         127.0.0.1:9890       0.0.0.0:*    users:(("cilium-agent",pid=1,fd=7))
tcp   LISTEN 0      1024        10.1.0.109:4240       0.0.0.0:*    users:(("cilium-agent",pid=1,fd=49))
tcp   LISTEN 0      1024         127.0.0.1:5353       0.0.0.0:*    users:(("cilium-agent",pid=1,fd=45))
tcp   LISTEN 0      1024                 *:4244             *:*    users:(("cilium-agent",pid=1,fd=50))
tcp   LISTEN 0      1024                 *:9091             *:*    users:(("cilium-agent",pid=1,fd=63))
tcp   LISTEN 0      1024                 *:9990             *:*    users:(("cilium-agent",pid=1,fd=8))


root@gke-us1-us1-default-2022-08-223cced9-cc8k:/home/cilium# iptables -tmangle -S | grep cilium-dns-egress
-A CILIUM_PRE_mangle -p tcp -m mark --mark 0x3b2c0200 -m comment --comment "cilium: TPROXY to host cilium-dns-egress proxy" -j TPROXY --on-port 11323 --on-ip 127.0.0.1 --tproxy-mark 0x200/0xffffffff
-A CILIUM_PRE_mangle -p udp -m mark --mark 0x3b2c0200 -m comment --comment "cilium: TPROXY to host cilium-dns-egress proxy" -j TPROXY --on-port 11323 --on-ip 127.0.0.1 --tproxy-mark 0x200/0xffffffff

root@gke-us1-us1-default-2022-08-223cced9-cc8k:/home/cilium# cat /var/run/cilium/state/proxy_ports_state.json
{"cilium-dns-egress":{"type":"dns","ingress":false,"port":11323}}
```
Is there some other way of working around this issue, or do we just need to wait for https://github.com/cilium/cilium/pull/36142 ?

**Comment by jrajahalme** [comment]
This should be fixed in 1.16.5 (labeled https://github.com/cilium/cilium/pull/36142 as a release blocker), scheduled for release mid-December.

I don't think there is anything you can do on 1.16.4 to prevent hitting this bug.
